Add Stable Diffusion 2.1 example

attila-dusnoki-htec opened the pull request:
This example show how to use HuggingFace's [Stable Diffusion v2.1](https://huggingface.co/stabilityai/stable-diffusion-2-1) model with migraphx's python frontend.

kahmed10 commented on examples/diffusion/python_stable_diffusion_21/README.md:
````diff
@@ -0,0 +1,49 @@
+Install dependencies
+
+```bash
+pip install optimum[onnxruntime] transformers diffusers accelerate
````
can you make a requirements.txt?

kahmed10 commented on examples/diffusion/python_stable_diffusion_21/README.md:
````diff
@@ -0,0 +1,49 @@
+```
+*Note: The first run will compile the models and cache them to make subsequent runs faster.*
+
+The result should like this:
````
```suggestion
The result should look like this:
```

kahmed10 commented on examples/diffusion/python_stable_diffusion_21/README.md:
```diff
@@ -0,0 +1,49 @@
+
+## Console application
+
+To run console application, follow these steps below.
```
```suggestion
To run the console application, follow these steps below.
```

kahmed10 commented on examples/diffusion/python_stable_diffusion_21/README.md:
````diff
@@ -0,0 +1,49 @@
+```
+*Note: `models/sd21-onnx` will be used in the scripts.*
+
+Run text to image example with the following example prompt and seed:
````
```suggestion
Run text-to-image script with the following example prompt and seed:
```

kahmed10 commented on examples/diffusion/python_stable_diffusion_21/sd21.ipynb:
```diff
@@ -0,0 +1,532 @@
+   "source": [
+    "Now it is time to load these models with python.\n",
+    "\n",
+    "First, we make sure that MIGraphX is in the path."
```
```suggestion
    "First, we make sure that MIGraphX module is found in the python path."
```

kahmed10 commented on examples/diffusion/python_stable_diffusion_21/sd21.ipynb:
```diff
@@ -0,0 +1,532 @@
+    "        print(f\"Saving {name} model to mxr file...\")\n",
+    "        mgx.save(model, f\"{file}.mxr\", format=\"msgpack\")\n",
+    "    else:\n",
+    "        print(f\"No {name} model found. Please download it and re-try.\")\n",
```
```suggestion
    "        print(f\"No {name} model found. Please verify the path is correct and re-try, or re-download model.\")\n",
```

kahmed10 commented on examples/diffusion/python_stable_diffusion_21/sd21.ipynb:
```diff
@@ -0,0 +1,532 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "With that, we can load the models. This could take minutes."
```
```suggestion
    "With that, we can load the models. This could take several minutes."
```

kahmed10 commented on examples/diffusion/python_stable_diffusion_21/sd21.ipynb:
```diff
@@ -0,0 +1,532 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Scale and decode the image latents with vae."
```
```suggestion
    "Scale and decode the image latents with VAE."
```

kahmed10 commented:
@attila-dusnoki-htec this looks good. Some minor suggestions above. Would it be possible to add some sort of timing information as well? Such as time it took to run the whole pipeline.

causten commented:
Can you reduce the png in size?  Perhaps convert to a jpg?   It is currently 420 KB.   Since it will be part of the base repo, the smaller the better

attila-dusnoki-htec commented:
> Can you reduce the png in size? Perhaps convert to a jpg? It is currently 420 KB. Since it will be part of the base repo, the smaller the better

Sorry about that.
I changed it to jpeg, it is 29K. But i can remove it as well if it is still too large.

attila-dusnoki-htec commented:
> @attila-dusnoki-htec this looks good. Some minor suggestions above. Would it be possible to add some sort of timing information as well? Such as time it took to run the whole pipeline.

Thanks. I fixed the grammar and added time measurement for certain parts of the code.
The output currently looks like this:  [example_run.log](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/files/13472287/example_run.log)

attila-dusnoki-htec commented:
I ended up adding a `gradio` version as well:

![gradio_sd_mgx](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/assets/126579622/651339fe-2326-4c82-ab9b-cfee91a6ab26)
